extensions/control-plane/api/management-api/asset-api/src/main/java/org/eclipse/edc/connector/api/management/asset/AssetNewApiController.java

```diff
@@ -0,0 +1,190 @@
+import jakarta.ws.rs.PUT;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.PathParam;
+import jakarta.ws.rs.Produces;
+import org.eclipse.edc.api.model.IdResponseDto;
+import org.eclipse.edc.api.query.QuerySpecDto;
+import org.eclipse.edc.api.transformer.DtoTransformerRegistry;
+import org.eclipse.edc.connector.api.management.asset.model.AssetEntryNewDto;
+import org.eclipse.edc.connector.api.management.asset.model.AssetResponseNewDto;
+import org.eclipse.edc.connector.api.management.asset.model.AssetUpdateRequestDto;
+import org.eclipse.edc.connector.api.management.asset.model.AssetUpdateRequestWrapperDto;
+import org.eclipse.edc.connector.api.management.asset.model.DataAddressDto;
+import org.eclipse.edc.connector.spi.asset.AssetService;
+import org.eclipse.edc.jsonld.util.JsonLdUtil;
+import org.eclipse.edc.spi.EdcException;
+import org.eclipse.edc.spi.asset.DataAddressResolver;
+import org.eclipse.edc.spi.query.QuerySpec;
+import org.eclipse.edc.spi.result.Result;
+import org.eclipse.edc.spi.types.domain.DataAddress;
+import org.eclipse.edc.spi.types.domain.asset.Asset;
+import org.eclipse.edc.web.spi.exception.InvalidRequestException;
+import org.eclipse.edc.web.spi.exception.ObjectNotFoundException;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static jakarta.ws.rs.core.MediaType.APPLICATION_JSON;
+import static java.util.Optional.of;
+import static java.util.Optional.ofNullable;
+import static java.util.stream.Collectors.toList;
+import static org.eclipse.edc.web.spi.exception.ServiceResultHandler.exceptionMapper;
+
+@Consumes(APPLICATION_JSON)
+@Produces(APPLICATION_JSON)
+@Path("/v2/assets")
+public class AssetNewApiController implements AssetNewApi {
+    private final DtoTransformerRegistry transformerRegistry;
+    private final AssetService service;
+    private final DataAddressResolver dataAddressResolver;
+
+    public AssetNewApiController(AssetService service, DataAddressResolver dataAddressResolver, DtoTransformerRegistry transformerRegistry) {
+        this.transformerRegistry = transformerRegistry;
+        this.service = service;
+        this.dataAddressResolver = dataAddressResolver;
+    }
+
+    @POST
+    @Override
+    public IdResponseDto createAsset(@Valid AssetEntryNewDto assetEntryDto) {
+        var expandedJson = JsonLdUtil.expand(assetEntryDto.getAsset());
+        if (expandedJson.size() == 0) throw new InvalidRequestException("Invalid asset!");
+
+        var expandedJsonAsset = expandedJson.getJsonObject(0);
+        var assetResult = transformerRegistry.transform(expandedJsonAsset, Asset.class);
+        var dataAddressResult = transformerRegistry.transform(assetEntryDto.getDataAddress(), DataAddress.class);
+
+        if (assetResult.failed() || dataAddressResult.failed()) {
+            var errorMessages = Stream.concat(assetResult.getFailureMessages().stream(), dataAddressResult.getFailureMessages().stream());
```

Comment:
I think this could be achieved with `Result.merge`, if one of them fails, it concats the message and return a failed one.

Reply:
This only works with results that have the same type parameter, which isn't the case here. We may to get this to work without the type constraint, but I really don't want to increase the PR surface even more.